File: zh-Hant/codes/csharp/chapter_graph/graph_adjacency_matrix.cs
/**
 * File: graph_adjacency_matrix.cs
 * Created Time: 2023-02-06
 * Author: zjkung1123 (zjkung1123@gmail.com)
 */

namespace hello_algo.chapter_graph;

/* 基於鄰接矩陣實現的無向圖類別 */
class GraphAdjMat {
    List<int> vertices;     // 頂點串列，元素代表“頂點值”，索引代表“頂點索引”
    List<List<int>> adjMat; // 鄰接矩陣，行列索引對應“頂點索引”

    /* 建構子 */
    public GraphAdjMat(int[] vertices, int[][] edges) {
        this.vertices = [];
        this.adjMat = [];
        // 新增頂點
        foreach (int val in vertices) {
            AddVertex(val);
        }
        // 新增邊
        // 請注意，edges 元素代表頂點索引，即對應 vertices 元素索引
        foreach (int[] e in edges) {
            AddEdge(e[0], e[1]);
        }
    }

    /* 獲取頂點數量 */
    int Size() {
        return vertices.Count;
    }

    /* 新增頂點 */
    public void AddVertex(int val) {
        int n = Size();
        // 向頂點串列中新增新頂點的值
        vertices.Add(val);
        // 在鄰接矩陣中新增一行
        List<int> newRow = new(n);
        for (int j = 0; j < n; j++) {
            newRow.Add(0);
        }
        adjMat.Add(newRow);
        // 在鄰接矩陣中新增一列
        foreach (List<int> row in adjMat) {
            row.Add(0);
        }
    }

    /* 刪除頂點 */
    public void RemoveVertex(int index) {
        if (index >= Size())
            throw new IndexOutOfRangeException();
        // 在頂點串列中移除索引 index 的頂點
        vertices.RemoveAt(index);
        // 在鄰接矩陣中刪除索引 index 的行
        adjMat.RemoveAt(index);
        // 在鄰接矩陣中刪除索引 index 的列
        foreach (List<int> row in adjMat) {
            row.RemoveAt(index);
        }
    }

    /* 新增邊 */
    // 參數 i, j 對應 vertices 元素索引
    public void AddEdge(int i, int j) {
        // 索引越界與相等處理
        if (i < 0 || j < 0 || i >= Size() || j >= Size() || i == j)
            throw new IndexOutOfRangeException();
        // 在無向圖中，鄰接矩陣關於主對角線對稱，即滿足 (i, j) == (j, i)
        adjMat[i][j] = 1;
        adjMat[j][i] = 1;
    }

    /* 刪除邊 */
    // 參數 i, j 對應 vertices 元素索引
    public void RemoveEdge(int i, int j) {
        // 索引越界與相等處理
        if (i < 0 || j < 0 || i >= Size() || j >= Size() || i == j)
            throw new IndexOutOfRangeException();
        adjMat[i][j] = 0;
        adjMat[j][i] = 0;
    }

    /* 列印鄰接矩陣 */
    public void Print() {
        Console.Write("頂點串列 = ");
        PrintUtil.PrintList(vertices);
        Console.WriteLine("鄰接矩陣 =");
        PrintUtil.PrintMatrix(adjMat);
    }
}

public class graph_adjacency_matrix {
    [Test]
    public void Test() {
        /* 初始化無向圖 */
        // 請注意，edges 元素代表頂點索引，即對應 vertices 元素索引
        int[] vertices = [1, 3, 2, 5, 4];
        int[][] edges = 
        [
            [0, 1],
            [0, 3],
            [1, 2],
            [2, 3],
            [2, 4],
            [3, 4]
        ];
        GraphAdjMat graph = new(vertices, edges);
        Console.WriteLine("\n初始化後，圖為");
        graph.Print();

        /* 新增邊 */
        // 頂點 1, 2 的索引分別為 0, 2
        graph.AddEdge(0, 2);
        Console.WriteLine("\n新增邊 1-2 後，圖為");
        graph.Print();

        /* 刪除邊 */
        // 頂點 1, 3 的索引分別為 0, 1
        graph.RemoveEdge(0, 1);
        Console.WriteLine("\n刪除邊 1-3 後，圖為");
        graph.Print();

        /* 新增頂點 */
        graph.AddVertex(6);
        Console.WriteLine("\n新增頂點 6 後，圖為");
        graph.Print();

        /* 刪除頂點 */
        // 頂點 3 的索引為 1
        graph.RemoveVertex(1);
        Console.WriteLine("\n刪除頂點 3 後，圖為");
        graph.Print();
    }
}

File: ru/codes/c/chapter_graph/graph_adjacency_matrix.c
/**
 * File: graph_adjacency_matrix.c
 * Created Time: 2023-07-06
 * Author: NI-SW (947743645@qq.com)
 */

#include "../utils/common.h"

// Предположить, что максимальное число вершин равно 100
#define MAX_SIZE 100

/* Структура неориентированного графа на основе матрицы смежности */
typedef struct {
    int vertices[MAX_SIZE];
    int adjMat[MAX_SIZE][MAX_SIZE];
    int size;
} GraphAdjMat;

/* Конструктор */
GraphAdjMat *newGraphAdjMat() {
    GraphAdjMat *graph = (GraphAdjMat *)malloc(sizeof(GraphAdjMat));
    graph->size = 0;
    for (int i = 0; i < MAX_SIZE; i++) {
        for (int j = 0; j < MAX_SIZE; j++) {
            graph->adjMat[i][j] = 0;
        }
    }
    return graph;
}

/* Деструктор */
void delGraphAdjMat(GraphAdjMat *graph) {
    free(graph);
}

/* Добавление вершины */
void addVertex(GraphAdjMat *graph, int val) {
    if (graph->size == MAX_SIZE) {
        fprintf(stderr, "Количество вершин графа уже достигло максимума\n");
        return;
    }
    // Добавить n-ю вершину и обнулить n-ю строку и столбец
    int n = graph->size;
    graph->vertices[n] = val;
    for (int i = 0; i <= n; i++) {
        graph->adjMat[n][i] = graph->adjMat[i][n] = 0;
    }
    graph->size++;
}

/* Удаление вершины */
void removeVertex(GraphAdjMat *graph, int index) {
    if (index < 0 || index >= graph->size) {
        fprintf(stderr, "индекс вершины выходит за границы\n");
        return;
    }
    // Удалить вершину с индексом index из списка вершин
    for (int i = index; i < graph->size - 1; i++) {
        graph->vertices[i] = graph->vertices[i + 1];
    }
    // Удалить строку с индексом index из матрицы смежности
    for (int i = index; i < graph->size - 1; i++) {
        for (int j = 0; j < graph->size; j++) {
            graph->adjMat[i][j] = graph->adjMat[i + 1][j];
        }
    }
    // Удалить столбец с индексом index из матрицы смежности
    for (int i = 0; i < graph->size; i++) {
        for (int j = index; j < graph->size - 1; j++) {
            graph->adjMat[i][j] = graph->adjMat[i][j + 1];
        }
    }
    graph->size--;
}

/* Добавление ребра */
// Параметры i и j соответствуют индексам элементов vertices
void addEdge(GraphAdjMat *graph, int i, int j) {
    if (i < 0 || j < 0 || i >= graph->size || j >= graph->size || i == j) {
        fprintf(stderr, "индексы ребра выходят за границы или совпадают\n");
        return;
    }
    graph->adjMat[i][j] = 1;
    graph->adjMat[j][i] = 1;
}

/* Удаление ребра */
// Параметры i и j соответствуют индексам элементов vertices
void removeEdge(GraphAdjMat *graph, int i, int j) {
    if (i < 0 || j < 0 || i >= graph->size || j >= graph->size || i == j) {
        fprintf(stderr, "индексы ребра выходят за границы или совпадают\n");
        return;
    }
    graph->adjMat[i][j] = 0;
    graph->adjMat[j][i] = 0;
}

/* Вывести матрицу смежности */
void printGraphAdjMat(GraphAdjMat *graph) {
    printf("Список вершин = ");
    printArray(graph->vertices, graph->size);
    printf("Матрица смежности =\n");
    for (int i = 0; i < graph->size; i++) {
        printArray(graph->adjMat[i], graph->size);
    }
}

/* Driver Code */
int main() {
    // Инициализация неориентированного графа
    GraphAdjMat *graph = newGraphAdjMat();
    int vertices[] = {1, 3, 2, 5, 4};
    for (int i = 0; i < 5; i++) {
        addVertex(graph, vertices[i]);
    }
    int edges[][2] = {{0, 1}, {0, 3}, {1, 2}, {2, 3}, {2, 4}, {3, 4}};
    for (int i = 0; i < 6; i++) {
        addEdge(graph, edges[i][0], edges[i][1]);
    }
    printf("\nПосле инициализации граф имеет вид\n");
    printGraphAdjMat(graph);

    /* Добавление ребра */
    // Индексы вершин 1 и 2 равны 0 и 2 соответственно
    addEdge(graph, 0, 2);
    printf("\nПосле добавления ребра 1-2 граф имеет вид\n");
    printGraphAdjMat(graph);

    /* Удаление ребра */
    // Индексы вершин 1 и 3 равны 0 и 1 соответственно
    removeEdge(graph, 0, 1);
    printf("\nПосле удаления ребра 1-3 граф имеет вид\n");
    printGraphAdjMat(graph);

    /* Добавление вершины */
    addVertex(graph, 6);
    printf("\nПосле добавления вершины 6 граф имеет вид\n");
    printGraphAdjMat(graph);

    /* Удаление вершины */
    // Индекс вершины 3 равен 1
    removeVertex(graph, 1);
    printf("\nПосле удаления вершины 3 граф имеет вид\n");
    printGraphAdjMat(graph);

    // Освободить память
    delGraphAdjMat(graph);

    return 0;
}
